File: scripts/linkcheck.py
# -*- coding: utf-8 -*-
"""linkcheck — кожне посилання має ціль або стаб «в розробці», інакше exit 1.

Перевіряє в усіх .md книги embedded:
  • локальні .md-лінки  ](path.md)  — файл-ціль мусить існувати;
  • крос-книжкові book:<id>/<slug>… — id мусить бути відомою книгою
    (рушій сам показує стаб «в розробці», якщо тема ще не написана);
  • зображення ](path.svg/png) — файл мусить існувати.
Друкує всі нерозв'язані лінки згруповано й завершується кодом 1, якщо такі є.

Запуск:  python embedded/_tools/linkcheck.py
"""
import io, os, re, sys, posixpath

ROOT = r"E:\develop\courses\embedded"
KNOWN_BOOKS = {"electronics", "chem", "math", "components"}
LINK_RE = re.compile(r'\]\(([^)]+?)\)')

missing_md = []
bad_book = []
missing_img = []
queued_md = []
checked = 0

# базові імена вставок/тем, що стоять у чергах _status.md — це легітимний стаб «в розробці»
QUEUED = set()
for dp, dn, fns in os.walk(ROOT):
    if '_status.md' in fns:
        try:
            t = io.open(os.path.join(dp, '_status.md'), encoding='utf-8').read()
        except Exception:
            t = ''
        for m in re.finditer(r'`([^`]+\.md)`', t):
            QUEUED.add(m.group(1).split('/')[-1])

for dp, dn, fns in os.walk(ROOT):
    for fn in fns:
        if not fn.endswith('.md'):
            continue
        full = os.path.join(dp, fn)
        rel = os.path.relpath(full, ROOT).replace('\\', '/')
        try:
            txt = io.open(full, encoding='utf-8').read()
        except Exception:
            continue
        base = posixpath.dirname(rel)
        for m in LINK_RE.finditer(txt):
            href = m.group(1).strip()
            if href.startswith('#') or href.startswith('mailto:') or href.startswith('tel:'):
                continue
            if '<' in href or '>' in href or '...' in href:   # шаблон-приклад у документації, не лінк
                continue
            hi = href.find('#')
            core = href[:hi] if hi >= 0 else href
            if core.startswith('book:'):
                checked += 1
                bid = core[len('book:'):].split('/', 1)[0]
                if bid not in KNOWN_BOOKS:
                    bad_book.append('%s  ->  %s' % (rel, href))
                continue
            if core.startswith('http://') or core.startswith('https://') or not core:
                continue
            if core.endswith('.md'):
                checked += 1
                tgt = posixpath.normpath(posixpath.join(base, core))
                if not os.path.isfile(os.path.join(ROOT, tgt.replace('/', os.sep))):
                    if posixpath.basename(core) in QUEUED:
                        queued_md.append('%s  ->  %s' % (rel, href))   # ціль у черзі _status.md
                    else:
                        missing_md.append('%s  ->  %s' % (rel, href))
            elif re.search(r'\.(svg|png|jpg|jpeg|gif|webp)$', core, re.I):
                checked += 1
                tgt = posixpath.normpath(posixpath.join(base, core))
                if not os.path.isfile(os.path.join(ROOT, tgt.replace('/', os.sep))):
                    missing_img.append('%s  ->  %s' % (rel, href))

def dump(title, items):
    print("\n=== %s (%d) ===" % (title, len(items)))
    for x in items:
        print("  ", x)

print("linkcheck: перевірено посилань:", checked)
dump("MISSING .md TARGETS (потрібен стаб або правка)", missing_md)
dump("UNKNOWN book: id", bad_book)
dump("MISSING IMAGE TARGETS", missing_img)
if queued_md:
    print("\n=== У ЧЕРЗІ — ціль ще не написана, але стоїть у _status.md (OK) (%d) ===" % len(queued_md))
    for x in queued_md:
        print("  ", x)

fail = len(missing_md) + len(bad_book) + len(missing_img)
if fail:
    print("\nFAIL:", fail, "нерозв'язаних посилань.")
    sys.exit(1)
print("\nOK: усі посилання мають ціль або стаб.")

File: scripts/linkcheck.js
/* scripts/linkcheck.js — перевірка посилань у дереві v7 (root/<вид>/<книга>/<тема>/).
   Перевіряє:
     • root:<книга>/<тема>[/<file>] — чи є така тема в маніфесті книги або курсу; status:"empty" = легітимний стаб.
     • зображення (/book/…/img/x.svg або відносні img/x.svg) — чи існує файл.
     • відносні .md-лінки — попереджає (за каноном крос-лінк має бути root:-попапом).
   Запуск: node scripts/linkcheck.js
   ========================================================================== */
"use strict";
const fs = require("fs");
const path = require("path");
const ROOT = path.resolve(__dirname, "..");

function read(p) { return fs.readFileSync(p, "utf8"); }
function exists(p) { try { fs.accessSync(p); return true; } catch (e) { return false; } }
function walk(dir, out) {
  if (!exists(dir)) return out;
  for (const e of fs.readdirSync(dir, { withFileTypes: true })) {
    const p = path.join(dir, e.name);
    if (e.isDirectory()) walk(p, out);
    else if (e.isFile() && e.name.endsWith(".md")) out.push(p);
  }
  return out;
}
function loadReg(globKey, file) {
  if (!exists(file)) return [];
  try { return new Function("window", read(file) + `\nreturn window.${globKey}||[];`)({}); } catch (e) { return []; }
}

/* Індекс тем усього корпусу: "<книга>/<тема>" → { dir, status, files }.
   v7: один розбір на весь тулінг — scripts/lib/manifest7.js. Книга й курс лежать
   в ОДНОМУ просторі імен, бо префікс лінка тепер теж один (topic:). */
const M7 = require("./lib/manifest7.js");
const TOPICS = new Map();
const GTOPICS = new Map();                      // лишаємо ім'я: на нього дивиться решта файла
for (const [bslug, meta] of M7.books()) {
  const bk = M7.loadBook(meta.bookDir);
  if (!bk) continue;
  for (const t of M7.allTopics(bk)) {
    if (!t.own) continue;
    const n = t.node;
    const files = new Set();
    const bs = (n.basic && n.basic.status) || "empty";
    const ds = (n.detailed && n.detailed.status) || "empty";
    if (bs !== "empty") files.add(t.slug + ".md");
    if (ds !== "empty") files.add(t.slug + "-d.md");
    for (const k of M7.INSERT_TYPES) for (const x of n[k] || []) files.add(typeof x === "string" ? x : x.file);
    /* Доступна читачу, якщо done БУДЬ-ЯКА версія. У v6 дивилися лише на basic, і в v7
       це дало 48 тис. фальшивих «стабів»: тема з basic:empty + detailed:done — норма. */
    const avail = (bs === "done" || ds === "done") ? "done" : (ds !== "empty" ? ds : bs);
    TOPICS.set(bslug + "/" + t.slug, { dir: path.join(meta.bookDir, t.slug), status: avail, files, group: t.group });
  }
}

const reLink = /(!?)\[[^\]]*\]\(([^)\s]+)(?:\s+"[^"]*")?\)/g;
const broken = [], warnMd = [], stubs = [], stale = [];
const mdFiles = walk(path.join(ROOT, "root"), []);

for (const f of mdFiles) {
  const rel = path.relative(ROOT, f);
  const txt = read(f);
  let m;
  while ((m = reLink.exec(txt))) {
    const isImg = m[1] === "!";
    let href = m[2];
    if (/^(https?:|mailto:|tel:|#)/i.test(href)) continue;
    href = href.split("#")[0]; if (!href) continue;

    if (/^root:/i.test(href)) {
      /* Резолв за PLAN §2.3: довга форма МІСТИТЬ коротку, тож беремо КНИГУ з першого
         сегмента і ТЕМУ з останнього змістовного — і коротка `книга/тема`, і довга
         `книга/група/розділ/тема` читаються одним правилом. Середина довгої — не адреса,
         а перевірка: розійшлася з маніфестом → адреса застаріла, а не бита.            */
      const segs = href.replace(/^root:/i, "").split("/").filter(Boolean);
      const book = segs[0] || "";
      const last = segs[segs.length - 1] || "";
      const isVer = /\.md$/i.test(last) || last === "detail" || last === "basic";
      const slug = (isVer ? segs[segs.length - 2] : last) || "";
      const ver = isVer ? last : null;
      const key = book + "/" + slug;
      const t = TOPICS.get(key) || GTOPICS.get(key);        // книга й курс — один простір імен
      if (!t) { broken.push(`${rel}: root:${segs.join("/")} — теми нема в жодному маніфесті`); continue; }
      /* Середні сегменти (група, розділ) — довідкові. Розбіжність із маніфестом ловимо
         окремим, мʼяким вироком: лінк робочий, але показує стару розкладку. */
      const mid = segs.slice(1, isVer ? segs.length - 2 : segs.length - 1);
      if (mid.length && t.group && mid[0] !== t.group) stale.push(`${rel}: root:${segs.join("/")} — група в адресі «${mid[0]}», у маніфесті «${t.group}»`);
      if (ver) {
        const want = ver === "detail" ? slug + "-d.md" : ver === "basic" ? slug + ".md" : ver;
        if (!t.files.has(want) && !exists(path.join(t.dir, want))) broken.push(`${rel}: root:${segs.join("/")} — файла нема`);
      } else if (t.status !== "done") stubs.push(`${rel}: root:${key} (${t.status}-стаб)`);
      continue;
    }
    // шлях від кореня репо (зображення/файл)
    const target = href.charAt(0) === "/" ? path.join(ROOT, href) : path.join(path.dirname(f), href);
    if (isImg || /\.(svg|png|jpg|jpeg|gif|webp)$/i.test(href)) {
      if (!exists(target)) broken.push(`${rel}: зображення ${href} — нема`);
    } else if (/\.md$/i.test(href)) {
      warnMd.push(`${rel}: відносний .md-лінк ${href} — за каноном має бути root:-попап`);
      if (!exists(target)) broken.push(`${rel}: .md-ціль ${href} — нема`);
    }
  }
}

console.log(`Перевірено .md: ${mdFiles.length} · тем у маніфестах: ${TOPICS.size}`);
console.log(`\n=== БИТІ (${broken.length}) ===`); broken.slice(0, 200).forEach((s) => console.log("  ✗ " + s));
console.log(`\n=== АДРЕСА ЗАСТАРІЛА (лінк робочий, показує стару розкладку) (${stale.length}) ===`); stale.slice(0, 30).forEach((s) => console.log("  ~ " + s));
console.log(`\n=== відносні .md-лінки, варто на root: (${warnMd.length}) ===`); warnMd.slice(0, 50).forEach((s) => console.log("  ⚠ " + s));
console.log(`\n=== лінки на empty-стаби (OK, ${stubs.length}) ===`);
process.exit(broken.length ? 1 : 0);
